### DISCUSSION TITLE
How to Break Down Assembly Purchases

### DISCUSSION THREAD

**Discussion by SchrodingersGat** [inventree/InvenTree#4722] [Q&A]
Refer to this issue - https://github.com/inventree/InvenTree/issues/1192

Basically, some parts may be purchased from a supplier in a "kit" which contains multiple parts we are interested in. The SuppliePart / SKU references a single item, but that item can be "broken down" into multiple component items.

This would be a useful feature to have, but has some complexities with regard to implementation. Following are some ways we *might* be able to implement it (but there may well be alternative solutions which are superior)

## Option 1: Assembly

Create a new internal part, marked as an [assembly](https://docs.inventree.org/en/latest/part/part/#assembly), which has a bill of materials. Define the bill of materials as the "contents" of the "kit".  Also define the part as "purchaseable" so it can be received against purchase orders

When receiving an item again a purchase order which is an "assembly" the user has the option to break it down into component items

### Advantages

- Makes use of existing data structures (Bills of Materials)

### Problems

- No way of associating purchase price with individual items (if this even makes sense)

## Option 2: Part Kit

Define a new type of part which is a "kit" - it functions much like an assembly in that it has a bill of materials, but instead of parts being "consumed" into the assembly, they are "gathered".

This may be a subtle distinction, and I'd have to think about this further

## Option 3: ???

Looking for feedback on alternative ideas, or insight into how other similar software implements this

**Comment by miggland** [comment]
I think this is a great idea. Some thoughts:

- There will never be an automatic way of associating the purchase price with individual items, regardless of which technical solution InvenTree implements, which is always correct. If this is wanted, it would have to be somehow defined by a user.
- This would be useful not only when buying sets, but also when selling them. We sell "kits" of multiple items, which to the customer has only one SKU. At the moment, this item has a BoM and I make a BuildOrder when required. It's a few extra steps when handling the order, but doable. Doesn't happen often at the moment, but if it did, some kind of fast-build would be useful.
- The reverse process would be for buying - in other words, what is missing is a "reversibility" of the build process in that individual parts are put in stock, and the assembled item is removed from stock.

I think the difficulties start when considering how to speed up, and asking when the assembly/disassembly should be carried out.
- When selling:
  - as soon as all required parts are in stock?
  - whenever someone in the warehouse puts items in boxes?
  - when adding a line to a SO?
  - when allocating the item to a shipment?
-  When buying:
   - When receiving the item?
   - Later on, for example when someone unpacks items?

Which answer is best may vary depending on workflows. Besides adding a the opposite of a BuildOrder, making the process quick for such sets would be helpful. But that raises questions as to where to take items from/put them (StockLocation).

**Comment by westinpigott** [comment]
This would be very helpful.   I think about it as a more complex BOM process, where we treat a build order instead as a work order.  We can specify the inputs  3X, 2Y, 7Z and the outputs  2H, 5G, 1J.  This way a work order can be any level of complexity. 

BOM would probably be better changed to a separate entity where it is a "Work Order Definition".

Adding the ability to add in labor cost/time as well as machine cost/time would be useful as well.

For cost flows, I would think we would have a ratio assigned to the outputs.    We can default it where (above example) 2+5+1 = 8, so each gets 100/8 = 12.5% of the total cost.  The user could change the cost allocation ratios in the Work Order Definition as well as long as it adds up to 100%.

**Comment by cpknight** [comment]
I keep coming up against this issue where the parts in my instance are for "socket wires" and "plug wires" but most suppliers sell these wires as socket/plug pair wires. I tend to think that a supplier part doesn't necessarily need to have a 1:1 relation with a base part type - more flexibility could allow users to define a supplier part as referring to 1 base part, 2 base parts, etc. 

It would also be nice to be able to setup supplier parts without necessarily having a base part in the system. 

When receiving purchase order line items, the logic in InvenTree would have to deal with two additional scenarios:

1. If no base parts are associated with the supplier part/PO line item, then no stock item will be created when then the line item is received (probably a user message should appear advising as such). 

2. If more than one base parts are associated with the supplier part/PO line item, then multiple stock items need to be created for the multiple base parts. 

The "add supplier part" functionality would also need to be updated to allow zero, one, or more than one "base part" to be associated with the supplier part. 

The underlying data schema will also be affected. 

Overall, however, I think the flexibility in being about to map "suppler parts" to "base parts" in a less rigid relationship would benefit InvenTree greatly - too many real-world scenarios are otherwise adversely affected by the assumption of a 1:1 relationship between what suppliers sell as a "unit" and what manufacturers consider as unique part "units".

**Reply by SchrodingersGat** [reply]
I think that your suggestions make sense - although the work required to ensure that all of the edge cases here would be complex.

**Comment by Carabela-del-Cielo** [comment]
I believe I'm hitting a similar issue when it comes to defining a salable product.  I'm newer to InvenTree and might have just missed the intended workflow too.  I was about to put in a feature request but this thread seems to be about what I am hitting.

As the example, I have:
- Keyboard part; assembly with SN
- Mouse part; assembly also with SN

and then packaging.

## Thoughts on Solutions

My sense is that in addition to "Salable", "Salable - Bundle" is also needed (or Kitted / Packaged / whatever).    "Bundle" is my term for this ready-to-sell product, but there may be a better one.  This extra designation might be all that is needed, but there are at least two ways to interpret that addition:
1. A Part now has multiple BOMs: one to build it, the others define the Bundles it can be sold as
	a. So it might manifest as a "keyboard" packaged individually, or a keyboard/mouse combo
	b. Basically, line items in a SO can be the 'bare' products: keyboards, mice
		1. Part selection for the line item would include selecting the bundle which pulls in all associated packaging
	c. Line items to the customer can be shown directly as the individual Parts, and not "custom keyboard box" and "foam padding"
	d. I'm thinking this will be the more convoluted solution; since both keyboard and mouse will need to point to the same bundle option, for instance

**OR,** 

2. The Bundle is just a Salable Part where its BOM has the Part and all packaging necessary for it
	a. So now I create a variety of bundles of keyboards, mice/keyboard combos, etc.
	b. When making a sales order, filtering by salable and checkbox for 'bundle' could make this nice
	c. This would account for all packaging needed for said bundle
But its advantage is being more filterable for special cases: sales orders might only want to look at Bundles, RMAs and other workflows can treat the Keyboard as the product, but the Bundle is what's SOLD.  I think this also lines up with "Option 2: Part Kit" that SchrodingersGat talks about.

Following westinpigott: A bundle might also include "Extra BOM items" (similar to the extra line items in SO's), to itemize machine time and such [In fact...making this into its own section makes sense with any part.  Heck, it can take twenty minutes just to liberate a package of purchased components].

It makes sense to think about this as roles with definite tasks (Build order vs fulfilment tasks), but separating a 'fulfilment' tab for handling shipping doesn't seem necessary, but I haven't given it much thought.
Also to note, I had first thought that "Extra Line Items" might allow me to include various items like packaging which are 'unsalable', but it does not.  Could tuck that stuff there for instance.

### Additional concerns: 

1. Often a varied quantity can fit in a box: maybe 2-15 parts fit in a box when shipped, so ability to specify size will help, or calculate for instance, a sale of 38 parts ==> 2 boxes of 15, 1 box of 8
2. If the customer wants 4 keyboards, I'd be selling one "4-keyboard bundle".  I'd want to be able to enter the line item using bundle quantity.
3. RMAs are tricky right now.  I would want to select the returned part and not just force it to be a sold part.  Or I could work with the sold part to associate it to a line item; but then specify the subassembly part that is being returned/repaired.  Particularly for a serialized part I should be able to select the serialized part itself as returned, not the sold assembly/package it was sent in.  That goes for a bundle or installed piece too:  say an LED kit in the keyboard has a serial number, i should be able to do an RMA just for the LED kit and not have to select the bundled keyboard, as it forces now.

There's also a balance of too much data entry vs what is necessary to be efficient and not lose stuff.  A few workers and multiple projects going on can become a mess quickly.  Too close and too far from modeling reality are both problematic.

Not sure if I'm off track on this or if InvenTree already provides this workflow, but this happens to be where I'm stuck now.  And also - thanks for a great product as it is now!
